### PR TITLE
Add AI Summarization and added Enhancements to YNAB Memo Processing and Testing

### PR DESCRIPTION
**Pull request by omarshahine**
# Enhancements to YNAB Memo Processing and Testing

## Overview
This PR significantly improves the memo processing functionality, adds comprehensive testing, and introduces optional AI summarization for Amazon order memos.

### Core Improvements
- Enhanced memo truncation logic to better preserve important information
- Improved handling of markdown formatting in memos
- Refactored code for better maintainability and error handling
- Added robust URL extraction and normalization
- Improved handling of partial order warnings

### Testing Infrastructure
- Added comprehensive test suite for memo processing
- Implemented test fixtures for consistent test data
- Added tests for edge cases and error conditions
- Improved test coverage across the codebase
- Added mocking for external dependencies

### New Features
- Added optional AI summarization for Amazon order memos
- Implemented fallback to standard truncation when AI summarization is unavailable
- Added new environment variables for AI configuration:
  - `YNAB_USE_AI_SUMMARIZATION=true/false`
  - `OPENAI_API_KEY=your-openai-api-key`

### Code Quality
- Improved error handling
- Refactored complex functions into smaller, more maintainable pieces
- Improved logging and error messages

### Documentation
- Updated README.md with:
  - New environment variables
  - AI summarization feature description
  - Example of AI-summarized memos
  - Configuration instructions
  - Improved examples of memo formats

## Testing
- Added comprehensive test suite for memo processing
- Implemented mocking for external dependencies
- Added tests for edge cases and error conditions
- All tests passing with improved coverage:
  - `ynab_memo.py`: 72% coverage
  - `amazon_transactions.py`: 51% coverage
  - `ynab_transactions.py`: 37% coverage

## Notes
- AI summarization is an optional feature that can be enabled/disabled
- Core memo processing improvements benefit all users, regardless of AI feature usage
- Improved error handling and fallback behavior throughout
- Better test coverage and maintainability

### CODE REVIEW

File: pyproject.toml
Comment (WoosterInitiative) [inline pyproject.toml:1]: perhaps this can be added as an "extra"? that way it doesn't need to be installed for users that don't want it?

I don't feel super strongly about this, though

File: src/ynamazon/settings.py [hunk 1]
```diff
@@ -25,16 +26,32 @@ def _display(self) -> str:
...
+        env_prefix="",
+        case_sensitive=True,
+        extra="ignore"
+    )
 
     ynab_api_key: SecretApiKey
     ynab_budget_id: SecretBudgetId
     amazon_user: EmailStr
     amazon_password: SecretStr
+    openai_api_key: SecretStr | None = None
```
Comment (WoosterInitiative) [inline src/ynamazon/settings.py]: perhaps use `SecretApiKey` instead, only so that a few characters can be visible in logs, etc.

File: src/ynamazon/settings.py [hunk 2]
```diff
@@ -25,16 +26,32 @@ def _display(self) -> str:
...
 
     ynab_payee_name_to_be_processed: str = "Amazon - Needs Memo"
     ynab_payee_name_processing_completed: str = "Amazon"
     ynab_use_markdown: bool = False
+    use_ai_summarization: bool = False
+    suppress_partial_order_warning: bool = False
+
+    @field_validator("ynab_use_markdown", "use_ai_summarization", "suppress_partial_order_warning", mode="before")
```
Comment (WoosterInitiative) [inline src/ynamazon/settings.py]: I haven't checked this, but I don't think these need to be specifically validated. Pydantic _should_, IIRC, do this for us with any "truthy" value.

File: src/ynamazon/ynab_memo.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+# Constants
+YNAB_MEMO_LIMIT = 500  # YNAB's character limit for memos
+
+
+def generate_ai_summary(
+    items: list[str],
+    order_url: str,
+    order_total: Optional[str] = None,
+    transaction_amount: Optional[str] = None,
+    max_length: int = 500
```
Comment (WoosterInitiative) [inline src/ynamazon/ynab_memo.py]: set the default to `YNAB_MEMO_LIMIT` instead

File: src/ynamazon/ynab_memo.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+) -> str:
```
Comment (WoosterInitiative) [inline src/ynamazon/ynab_memo.py]: if `None` is a valid response, it needs to be a possible return type here

`... -> Union[str, None]`

File: src/ynamazon/ynab_memo.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        order_url: Amazon order URL
+        order_total: Total order amount (if different from transaction)
+        transaction_amount: Current transaction amount
+        max_length: Maximum allowed characters (default: 500)
+    
+    Returns:
+        A human-readable memo summarized by AI
+    """
+    # Check if OpenAI key is available
+    if not settings.openai_api_key.get_secret_value():
```
Comment (WoosterInitiative) [inline src/ynamazon/ynab_memo.py]:
```
if settings.open_api_key is None:
  # i'd prefer this raise an exception, perhaps a custom `MissingOpenAIAPIKey` exception?
  raise MissingOpenAIAPIKey("OpenAI API key not found")
```

This then is used with the assumption that `use_ai_summarization` was enabled for this to be called at all, so it's an exception for those to be in opposition. This could also be done with a `model_validator` at the settings level?

File: src/ynamazon/ynab_memo.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+    
+    try:
+        # Get the response from OpenAI
+        response = client.chat.completions.create(
+            model="gpt-4o-mini",
+            messages=[
+                {"role": "system", "content": AMAZON_SUMMARY_SYSTEM_PROMPT},
+                {"role": "user", "content": full_prompt}
+            ]
+        )
```
Comment (WoosterInitiative) [inline src/ynamazon/ynab_memo.py:74]: tighten up the try...except block. The most likely place for an exception to occur is during the `create` function.

```
try:
  response = client.chat.completions.create(...)
except openai.AuthenticationError as e:
  # again, I think I like a custom exception
  raise InvalidOpenAIAPIKey() from e
```

if there are other openai exceptions we want to catch, we can do that with additional `except` clauses

File: src/ynamazon/ynab_memo.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+        
+        summary = response.choices[0].message.content.strip()
```
Comment (WoosterInitiative) [inline src/ynamazon/ynab_memo.py]: because it is possible for `content` to be None, `strip()` could raise an exception. let's handle that

```
message_content = response.choices[0].message.content
if message_content is None:
  raise ValueError("No content returned from OpenAI.")
summary = message_content.strip()  # this is now safe because we know message_content has a value
```

File: src/ynamazon/ynab_memo.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        
+        # If summary is too long, truncate it
+        if len(summary) > available_space:
+            summary = summary[:available_space-3] + "..."
+        
+        memo += f"{summary}\n{order_url}"
+            
+        return memo
+        
+    except Exception as e:
```
Comment (WoosterInitiative) [inline src/ynamazon/ynab_memo.py:83]: un-indent everything that is no longer inside the try...except block and remove this final except

also note: it's generally frowned upon to catch the generic `Exception` because it catches *everything*

File: tests/ynab/test_memo_truncation.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+
+    # Test markdown URL
+    memo = "Order details: [Order #123-4567890-1234567](https://www.amazon.com/gp/your-account/order-details?orderID=123-4567890-1234567)"
+    assert extract_order_url(memo) == "https://www.amazon.com/gp/your-account/order-details?orderID=123-4567890-1234567"
+
+    # Test no URL
+    memo = "No URL here"
+    assert extract_order_url(memo) is None
+
+def test_generate_ai_summary_error_handling(mock_settings):
```
Comment (WoosterInitiative) [inline tests/ynab/test_memo_truncation.py:207]: this test is _*not*_ actually testing what it looks like it's testing, because of how exceptions are currently handled, when debugging this test line-by-line, you'll find that the failure is actually due to the invalid key.

instead, mock the `create` method so that we can control that without a valid key and then check the results

**Review by WoosterInitiative — CHANGES_REQUESTED** [review]
Raising exceptions when appropriate instead of swallowing them whole allows for easier debugging and more robust code overall.

File: src/ynamazon/settings.py [hunk 10]
```diff
@@ -25,16 +26,32 @@ def _display(self) -> str:
...
+    @classmethod
+    def convert_bool(cls, v: str | bool) -> bool:
+        if isinstance(v, bool):
+            return v
+        return v.lower() in ("true", "1", "t", "y", "yes")
```
Comment (WoosterInitiative) [inline src/ynamazon/settings.py:56]: add this as a model_validator to make sure that if `use_ai_summarization` is `True`, then `openai_api_key` is not `None`

```python
@model_validator(mode="after")
def validate_ai_settings(self):
  if self.use_ai_summarization:
    if self.openai_api_key is None:
      raise ValueError("OpenAI API key must be set when using AI summarization.")
  return self
```

File: src/ynamazon/settings.py (hunk 2 above)
Comment (WoosterInitiative) [reply src/ynamazon/settings.py]: I've checked, all of the values that are being checked against in this custom validator would already be handled by pydantic's default behavior. please remove this entire method

**Comment by omarshahine** [comment]
Thanks for the wonderful comments. I'll take care of all this tonight.

File: pyproject.toml
Comment (omarshahine) [reply pyproject.toml:1]: Done

File: src/ynamazon/ynab_memo.py (hunk 3 above)
Comment (omarshahine) [reply src/ynamazon/ynab_memo.py]: Done

File: src/ynamazon/ynab_memo.py (hunk 6 above)
Comment (omarshahine) [reply src/ynamazon/ynab_memo.py:74]: Done

File: src/ynamazon/ynab_memo.py (hunk 5 above)
Comment (omarshahine) [reply src/ynamazon/ynab_memo.py]: done

File: src/ynamazon/ynab_memo.py (hunk 8 above)
Comment (omarshahine) [reply src/ynamazon/ynab_memo.py:83]: done

**Comment by omarshahine** [comment]
All done with changes. Thank you for the thorough review!

File: src/ynamazon/settings.py (hunk 10 above)
Comment (WoosterInitiative) [reply src/ynamazon/settings.py:56]: can I ask why you didn't use the model validator instead of the field validator? there are two reasons I prefer it, though the first one could also be done with a field validator
1. we can use `mode="after"` so that we don't have to deal with raw inputs, the values have already been coerced into the appropriate types
2. we don't have to hope that `open_ai_key` is present in `info.data` or make sure that it is listed before `use_ai_summarization` for the validation to work

I generally prefer to use model validation when I need to check for the presence of an attribute or use the data from more than one field.

File: src/ynamazon/ynab_memo.py (hunk 5 above)
Comment (WoosterInitiative) [reply src/ynamazon/ynab_memo.py]: this will still raise an exception if the value is `None`. Better to simply use `if settings.openai_api_key is None:` and _then_ raise `MissingOpenAIAPIKey`

File: src/ynamazon/settings.py [hunk 11]
```diff
@@ -25,16 +27,39 @@ def _display(self) -> str:
...
     amazon_password: SecretStr
+    openai_api_key: SecretApiKey | None = None
 
     ynab_payee_name_to_be_processed: str = "Amazon - Needs Memo"
     ynab_payee_name_processing_completed: str = "Amazon"
     ynab_use_markdown: bool = False
+    use_ai_summarization: bool = False
+    suppress_partial_order_warning: bool = False
+
+    @classmethod
```
Comment (WoosterInitiative) [inline src/ynamazon/settings.py]: this is no longer being used and should be removed

Comment (WoosterInitiative) [reply src/ynamazon/settings.py]: to be clear, the entire `convert_bool` method should be removed, pydantic will handle this on its own

**Review by WoosterInitiative — CHANGES_REQUESTED** [review]
just a simple change to remove an unused method from the settings class

**Review by WoosterInitiative — APPROVED** [review]
